File: lintcode/remove-duplicates-from-sorted-array.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/remove-duplicates-from-sorted-array/

class Solution:
    """
    @param A: a list of integers
    @return an integer
    """
    def removeDuplicates(self, A):
        # write your code here
        if len(A) <= 1:
            return len(A)

        writer = 0
        reader = 1
        while reader < len(A):
            while reader < len(A) and A[reader] == A[writer]:
                reader += 1
            if reader < len(A):
                writer += 1
                A[writer] = A[reader]

        return writer + 1

if __name__ == '__main__':
    s = Solution()
    array = [1,1,2]
    length = s.removeDuplicates(array)
    print length, array

File: leetcode/remove-duplicates-from-sorted-array.cc
// Question: https://leetcode.com/problems/remove-duplicates-from-sorted-array/

#include "headers.h"

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        int i = 0;
        int j = 0;
        while (j < nums.size()) {
            if (nums[i] != nums[j]) {
                ++i;
                swap(nums[i], nums[j]);
            }
            ++j;
        }
        return min((int)nums.size(), i+1);
    }

private:
    void swap(int& x, int& y) {
        int t = x;
        x = y;
        y = t;
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<int> nums{1,1,2,2,2,3,3,4,5,5,6};
    cout << s.removeDuplicates(nums) << endl;
    return 0;
}
